### PR TITLE
refactor(hydroflow_plus)!: disentangle instantiated nodes from locations

### PR DESCRIPTION
**Pull request by shadaj**
---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/hydro-project/hydroflow/pull/1395).
* #1398
* __->__ #1395

### CODE REVIEW

**Comment by shadaj** [comment]
Code is complete, need to get docs synced up

**Comment by shadaj** [comment]
The next PR in the stack also needs docs updates, so I'll just do it there.

File: hydroflow_plus_test/src/cluster/map_reduce.rs [hunk 1]
```diff
@@ -1,19 +1,18 @@
...
-    cluster_spec: impl ClusterSpec<'a, D>,
-) -> (D::Process, D::Cluster) {
-    let process = flow.process(process_spec);
-    let cluster = flow.cluster(cluster_spec);
+pub struct Leader {}
+pub struct Worker {}
+
+pub fn map_reduce(flow: &FlowBuilder) -> (Process<Leader>, Cluster<Worker>) {
+    let process = flow.process();
+    let cluster = flow.cluster();
```
Comment (MingweiSamuel) [inline hydroflow_plus_test/src/cluster/map_reduce.rs:9]: At some point we could make the struct declarations into a macro

**Review by MingweiSamuel — APPROVED** [review]
Didn't see anything suspicious

**Pull request merged**